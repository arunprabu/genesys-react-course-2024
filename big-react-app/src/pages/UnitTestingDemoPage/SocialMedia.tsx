import { useState } from "react";

interface SocialMediaProps {
  name: string;
  followersCount: string;
}

const SocialMedia: React.FC<SocialMediaProps> = ({ name, followersCount }) => {
  const [countryName, setCountryName] = useState("India");

  return (
    <div>
      <h2>Social Media Info!!</h2>
      <h3 className="text-center">
        Demo of Testing Props, States, Events, JSX, Styles, Snapshot Testing
      </h3>

      <p>Top Social Media Platform of Genesys</p>
      <p data-testid="followersDetails" style={{ backgroundColor: "#ffff00" }}>
        {name} with {followersCount} followers
      </p>

      <input
        type="text"
        placeholder="Enter Country Name"
        value={countryName}
        onChange={(event) => {
          setCountryName(event.target.value);
        }}
      />
      <p data-testid="websiteInfo">Visit Genesys {countryName} Website</p>
    </div>
  );
};

export default SocialMedia;
