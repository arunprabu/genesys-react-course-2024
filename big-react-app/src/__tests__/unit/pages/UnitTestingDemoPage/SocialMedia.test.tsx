// AAA
// Arrange
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';
import SocialMedia from "../../../../pages/UnitTestingDemoPage/SocialMedia";

// Group of related tests == TEST SUITE
describe("SocialMedia", () => {
  // Act
  // TESTING JSX
  it("has Social Media in heading", () => {
    render(<SocialMedia name="Instagram" followersCount="1.2M+" />);
    const headingElement = screen.getByText(/social media info/i); // using regex here
    // Assert
    expect(headingElement).toBeDefined();
  });

  // Testing Props
  it("receives the props name, followersCount and displays in paragraph", () => {
    render(<SocialMedia name="Instagram" followersCount="700K+" />);

    const followersDetailsElement = screen.getByTestId("followersDetails");
    // Instagram with 700K+ followers
    expect(followersDetailsElement.textContent).toBe(
      "Instagram with 700K+ followers"
    );
  });

  // Testing whether the form element is present or not
  it("should have input field with placeholder 'Enter Country Name'", () => {
    render(<SocialMedia name="YouTube" followersCount="500K+" />);
    const countryInput = screen.getByPlaceholderText("Enter Country Name");
    expect(countryInput).toBeTruthy();
  });

  // Testing input field's default value 
  it("should have input field with default value India", () => {
    render(<SocialMedia name="YouTube" followersCount="500K+" />);
    // find the element in which you have on change event
    const countryInput: HTMLInputElement = screen.getByRole("textbox");

    // let's test whether the input has default text
    expect(countryInput.value).toBe("India");
  });

  // Testing Events and States
  it('should update country name whenever the input field value is changed', () => {
    render(<SocialMedia name="YouTube" followersCount="500K+" />);
    // find the element in which you have on change event
    const countryInput: HTMLInputElement = screen.getByRole("textbox");

    // through program let's trigger the event -- because we can't use our input devices such as mouse, keyboard
    // having mock event obj
    // Triggering / emitting / firing event through program
    fireEvent.change(countryInput, {
      target: {
        value: 'Singapore'
      }
    });
    // assert 
    expect(countryInput.value).toBe('Singapore');
     
    const websiteInfoElement = screen.getByTestId('websiteInfo');
    expect(websiteInfoElement.textContent).toBe('Visit Genesys Singapore Website');
  });

  // Testing inline Styles
  it("displays social media followers count with background color yellow ", () => {
    render(<SocialMedia name="YouTube" followersCount="500K+" />);
    const followersDetailsElement = screen.getByTestId("followersDetails");
    expect(followersDetailsElement).toHaveStyle("background-color: #ffff00");
  });

  // TODO: Testing whether the element element has some css class

  // Snapshot Testing 
  it('has right SocialMedia component snapshot', () => {
    // to Take snapshot we need -- npm i react-test-renderer 
    // [RECOMMENDED: Take snapshot with all possible props as well props children]
    const snapshotJson = renderer.create(<SocialMedia name="YouTube" followersCount="500K+" />);
    // assert
    expect(snapshotJson).toMatchSnapshot();
  });
});
