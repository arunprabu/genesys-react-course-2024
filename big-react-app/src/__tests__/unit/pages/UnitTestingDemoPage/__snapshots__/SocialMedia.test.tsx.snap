// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialMedia has right SocialMedia component snapshot 1`] = `
<div>
  <h2>
    Social Media Info!!
  </h2>
  <h3
    className="text-center"
  >
    Demo of Testing Props, States, Events, JSX, Styles, Snapshot Testing
  </h3>
  <p>
    Top Social Media Platform of Genesys
  </p>
  <p
    data-testid="followersDetails"
    style={
      {
        "backgroundColor": "#ffff00",
      }
    }
  >
    YouTube
     with 
    500K+
     followers
  </p>
  <input
    onChange={[Function]}
    placeholder="Enter Country Name"
    type="text"
    value="India"
  />
  <p
    data-testid="websiteInfo"
  >
    Visit Genesys 
    India
     Website
  </p>
</div>
`;
